src/c/tb_client/context.zig

```diff
@@ -13,67 +30,278 @@ pub const ContextImplementation = struct {
-                allocator,
-                cluster_id,
-                addresses,
-                num_packets,
-                Context.on_completion,
+                context,
             ) catch |err| {
-                allocator.destroy(context);
-                return switch (err) {
-                    error.Unexpected => .unexpected,
-                    error.OutOfMemory => .out_of_memory,
-                    error.InvalidAddress => .invalid_address,
-                    error.SystemResources => .system_resources,
-                    error.NetworkSubsystemFailed => .network_subsystem,
+                log.err("failed to initalize thread: {}", .{err});
+                return err;
+            };
+            errdefer context.thread.deinit(context.allocator);
+
+            return context;
+        }
+
+        pub fn destroy(self: *Context) void {
+            self.thread.deinit();
+
+            self.client.deinit(self.allocator);
+            self.message_bus.deinit(self.allocator);
+            self.message_pool.deinit(self.allocator);
+            self.io.deinit();
+
+            self.allocator.free(self.addresses);
+            self.allocator.free(self.packets);
+            self.allocator.destroy(self);
+        }
+
+        pub fn tick(self: *Context) void {
+            self.client.tick();
+        }
+
+        pub fn run(self: *Context) void {
+            while (!self.thread.signal.is_shutdown()) {
+                self.client.tick();
+                self.io.run_for_ns(config.tick_ms * std.time.ns_per_ms) catch |err| {
+                    log.err("IO.run() failed with {}", .{err});
+                    @panic("IO.run() failed.");
                 };
+            }
+        }
+
+        pub fn request(self: *Context, packet: *Packet) void {
+            const message = self.message_pool.get_message();
+            defer self.message_pool.unref(message);
+            self.available_messages -= 1;
+
+            // Get the size of each request structure in the packet.data:
+            const request_size: usize = operation_size_of(packet.operation) orelse {
+                return self.on_complete(packet, error.InvalidOperation);
             };
 
-            var list = tb_packet_list_t{};
-            for (context.thread.packets) |*packet| {
-                list.push(tb_packet_list_t.from(packet));
+            // Make sure the packet.data size is correct:
+            const readable = packet.data[0..packet.data_size];
+            if (readable.len == 0 or readable.len % request_size != 0) {
+                return self.on_complete(packet, error.InvalidDataSize);
             }
 
-            out_packets.* = list;
-            return .success;
+            // Make sure the packet.data wouldn't overflow a message:
+            const writable = message.buffer[@sizeOf(Header)..];
+            if (readable.len > writable.len) {
+                return self.on_complete(packet, error.TooMuchData);
+            }
```

Comment:
The client does not expose the max message size (since it is configured at build time), so this error is returned if the application tries to submit more data than it is supported.

Writing this, I just realized that the client **must be** compiled with the same parameter the cluster is operating ... It is something tricky to do, especially when we distribute pre-built packages.

Reply:
Thanks! Yeah, if the application can hit this it should definitely be a soft error.

> Writing this, I just realized that the client must be compiled with the same parameter the cluster is operating ... It is something tricky to do, especially when we distribute pre-built packages.

Yeah, I have a few ideas around how to improve this. (These are in planning, not to be implemented yet):
- Eventually clients will provide a higher-level interface that handles batching on behalf of the application.
  - (Though we will still support the lower-level manual-batching interface.)
- Clients will expose the maximum number of events per operation (e.g. the maximum number of accounts they can create in a single batch). This will be partially negotiated with the cluster.